### ISSUE TITLE
Fluture with `esm`

### ISSUE THREAD

**Issue by henriklundgren** [fluture-js/Fluture#274]
I am unable to use Fluture with [`esm`](https://github.com/standard-things/esm/).

```js
// @file index.js
import {Future} from 'fluture';
```

```cli
$ node -r esm index.js
SyntaxError:
```

**Comment by yeldiRium** [comment]
Have you tried `import Future form 'fluture';`?

I haven't used it with esm yet, but I'm using `--experimental-modules` and it works fine.

**Comment by henriklundgren** [comment]
@yeldiRium Thanks, but it seems that it's not exporting anything, named nor default.

**Comment by Avaq** [comment]
Thanks for the report. I've reproduced it:

```console
$ cd `mktemp -d`
$ npm i fluture esm
+ fluture@9.0.1
+ esm@3.0.80
added 8 packages from 3 contributors and audited 12 packages in 1.359s
found 0 vulnerabilities
$ echo 'import {reject} from "fluture";' > index.js
$ node -r esm index.js 
file:///tmp/tmp.CRUBSFiSqW/index.js:1
SyntaxError: Missing export name 'reject' in ES module: file:///tmp/tmp.CRUBSFiSqW/node_modules/fluture/index.js
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:700:10)
```

The reason is that the wrong file is being loaded (`index.js`, as opposed to `index.mjs.js`).

The `package.json` specifies the `.mjs.js` file as being the ES module:

https://github.com/fluture-js/Fluture/blob/7c498a3e138c9905b6cb3c670932554ff65311c6/package.json#L7

I'll investigate further.

**Comment by Avaq** [comment]
I believe this happens because the "module" field in package.json is a "standard" set by the various build tools out there to distinguish between modular JavaScript and ESM. Node, however, decided to go with file extension to make the distinction.

So when loading a dependency with Node's `--experimental-modules` (or [`esm`](https://github.com/standard-things/esm/) by extension, because it follows Node), then it will try to load a `index.mjs`, and if it doesn't exist, fall back to `index.js`.

This is why `index.js` is being loaded, and not `index.mjs.js`.

This issue is loosely related to the ["Upgrade to ESM"](https://github.com/fluture-js/Fluture/projects/3#card-11238906) item on the Roadmap. Once I would fix the file extensions on the modular JavaScript files, Node and [`esm`](https://github.com/standard-things/esm/) would be able to identify them, and load them appropriately.

The reason I haven't done this yet is because many tools are not ready for the change. Babel, for example, only *just* started acknowledging the existence of the `.mjs` file extension.

**Comment by Avaq** [comment]
@henriklundgren I think you might be able to work around it for now by setting the `"mode"` [option](https://github.com/standard-things/esm#options) of esm to `"auto"`.

**Comment by Avaq** [comment]
@henriklundgren I'm working on a fix in #276 :)

**Comment by Avaq** [comment]
Seems like that branch fixes the issue.

```console
$ npm i fluture-js/fluture#avaq/mjs esm
+ esm@3.0.80
+ fluture@9.0.2
added 8 packages from 3 contributors and audited 12 packages in 4.697s
found 0 vulnerabilities

$ echo 'import {resolve} from "fluture"; resolve(42).value(console.log);' > index.js

$ node -r esm index.js 
42

$ cp index.js index.mjs 
$ node --experimental-modules --no-warnings index.mjs 
42
```

I will be releasing it soon along with some additional minor breaking changes. Until then, @henriklundgren, you can install Fluture with `npm install --save fluture-js/fluture#0aa28b3` to get the esm-enabled version.
